src/operator/src/statement.rs

```diff
@@ -120,7 +121,15 @@ impl StatementExecutor {
     ) -> Result<Output> {
         let _slow_query_timer = self.stats.start_slow_query_timer(stmt.clone());
         match stmt {
-            QueryStatement::Sql(stmt) => self.execute_sql(stmt, query_ctx).await,
+            QueryStatement::Sql(stmt) => {
+                if let Some(timeout) = query_ctx.query_timeout() {
+                    return tokio::time::timeout(timeout, self.execute_sql(stmt, query_ctx))
```

Comment:
`execute_sql` returns a stream (the output is a stream) and we may spend most time waiting on the stream.

We may need to add the timeout to the stream (But this is per-item timeout)
https://docs.rs/tokio-stream/latest/tokio_stream/trait.StreamExt.html#method.timeout

Reply:
This is a great catch I need to think more about where to apply the timeouts. After another review of docs, the implementation is tricker than i thought - [mysql](https://dev.mysql.com/doc/refman/8.4/en/server-system-variables.html#sysvar_max_execution_time ) query timeout only apply to the read statements. 

I will turn this pr to draft for now and add a second commit for applying timeout soon

Reply:
I have been thinking about how to apply per statement timeout in GreptimeDB.

Hopefully my understanding is correct with the life of a statement execution.
**execute**
(https://github.com/GreptimeTeam/greptimedb/blob/2f2b4b306cb480c1396f452f2c27dc2206705d38/src/frontend/src/instance.rs#L252C1-L295C14) create streams for each statement and the results are collected into a vec.
**result yielding**
postgres - the stream [yielding](https://github.com/sunng87/pgwire/blob/368e49ed5932de78deae24f20c3e806463b0506a/src/api/query.rs#L375C1-L380C6) at upstream pgwire package.
mysql - the stream [yielding](https://github.com/GreptimeTeam/greptimedb/blob/a889ea88ca595c184c14a3c65fe64027d9ed23a6/src/servers/src/mysql/writer.rs#L164C1-L182C22) at the writer.

Reply:
I thought about 3 additional approaches for the querytimeout.
1. attach timeouts at very beginning of query execution. i.e.  (add timeout to `on_query` method from `SimpleQueryHandler` for postgres).
pros: easy to implement, its also capture the query's entire life time.
cons: its wouldn't work on per statement timeout with bulk statements.
2. Only apply timeouts while waiting for the stream's results, attach timeout duration to the stream and use tokio select while yielding results from stream.
pros: easy to implement, can achieve per statement timeout
cons: may need to patch upstream pgwire
3. Alarm approach
- start alarm struct that waits for timeouts duration and send a query cancellation request after it wake up.
- when stream is created we could bind alarm and stream into the same lifetime.
cons: implement complexity



should we also update the slow query timer to account for the waiting time for stream records?
`let _slow_query_timer = self.stats.start_slow_query_timer(stmt.clone());`

Reply:
cc @evenyag @sunng87

Reply:
@lyang24 Sorry for the late reply. I've been quite busy in the past two weeks 😢 

> cons: its wouldn't work on per statement timeout with bulk statements.

Per statement timeout is enough.

I'd vote for approach 2 but also apply the timeout to the method call itself.

We can implement the timeout logic in the [plan_exec()](https://github.com/GreptimeTeam/greptimedb/blob/2f2b4b306cb480c1396f452f2c27dc2206705d38/src/operator/src/statement.rs#L383-L390).

> [mysql](https://dev.mysql.com/doc/refman/8.4/en/server-system-variables.html#sysvar_max_execution_time) query timeout only apply to the read statements.

We may check whether the query is a read query in `execute_sql()`.
https://github.com/GreptimeTeam/greptimedb/blob/2f2b4b306cb480c1396f452f2c27dc2206705d38/src/operator/src/statement.rs#L128-L132

We can wrap the implementation without a timeout into a method like `plan_exec_inner()` and attach the timeout to it if provided in `plan_exec()` first.

Then we can use `tokio select` and the `stream!` macro to implement a timeout stream that tracks the timeout in the statement level.

Once we get the `Vec<Output>` from `plan_exec_inner()`, we can attach the timeout stream to the output. The `attach_timer()` is an example.
https://github.com/GreptimeTeam/greptimedb/blob/2f2b4b306cb480c1396f452f2c27dc2206705d38/src/frontend/src/instance.rs#L365-L376

Reply:
thanks for the guidance I learned a lot and I have implemented a draft of second commit. I wonder if i could get another around of feedback/ suggestions on testing when you have time (no rush at all).